embedded_rtic/examples/sonars.rs

```diff
@@ -0,0 +1,440 @@
+//! Defines a simple distance measurement example.
+//!
+//! This example measures the distance to a nearby object, prefferably a wall
+//! using a sonar sensor. It then smooths the result over a few timestamps to
+//! avoid small peaks in the measured distance.
+
+#![no_main]
+#![no_std]
+#![feature(type_alias_impl_trait)]
+#![deny(clippy::all)]
+#![deny(warnings)]
+
+use test_app as _; // global logger + panicking-behavior + memory layout
+
+#[rtic::app(
+    device = nrf52840_hal::pac,
+    dispatchers = [RTC0,RTC1,RTC2]
+)]
+
+mod app {
+
+    use arraydeque::{behavior::Wrapping, ArrayDeque};
+    use defmt::{error, info};
+    use nrf52840_hal::{
+        clocks::Clocks, gpio::{self, Input, Output, Pin, PullDown, PushPull}, gpiote::*, ppi, prelude::*
+    };
+    use rtic_monotonics::{
+        nrf::timer::{fugit::ExtU64, Timer0 as Mono},
+        *,
+    };
+    use rtic_sync::{
+        channel::{Receiver, Sender},
+        make_channel,
+    };
+
+    /// The message queue capacity
+    const CAPACITY: usize = 5;
+
+    /// The duration type
+    type Instant = <Mono as rtic_monotonics::Monotonic>::Instant;
+
+    #[shared]
+    struct Shared {
+        gpiote: Gpiote,
+    }
+
+    // Local resources go here
+    #[local]
+    #[allow(dead_code)]
+    struct Local {
+        //Sonar 1
+        trig: Pin<Output<PushPull>>,
+        echo: Pin<Input<PullDown>>,
+        sender: Sender<'static, u32, CAPACITY>,
+        receiver: Receiver<'static, u32, CAPACITY>,
+        previous_time: Instant,
+
+        // Sonar 2
+        trig2: Pin<Output<PushPull>>, 
```

Comment:
This is not really needed for the example but when we write the real app we should use more legible names

```rust
trigger_front, tigger_left, trigger_right
```
etc etc